src/main/java/net/openhft/chronicle/core/io/VanillaReferenceCounted.java

```diff
@@ -175,6 +187,28 @@ public boolean unmonitored() {
         return unmonitored;
     }
 
+    public void addReferenceChangeListener(ReferenceChangeListener referenceChangeListener) {
+        referenceChangeListeners.add(referenceChangeListener);
+    }
+
+    public void removeReferenceChangeListener(ReferenceChangeListener referenceChangeListener) {
+        referenceChangeListeners.remove(referenceChangeListener);
+    }
+
+    private void callReferenceChangeListeners(ChangeType changeType, ReferenceOwner referenceOwner) {
+        //noinspection ForLoopReplaceableByForEach
+        for (int i = 0; i < referenceChangeListeners.size(); i++) {
+            ReferenceChangeListener listener;
+            try {
+                listener = referenceChangeListeners.get(i);
+            } catch (ArrayIndexOutOfBoundsException e) {
+                // This can happen if a listener is removed concurrently during the iteration
+                continue;
```

Comment:
I think this can happen, CopyOnWriteArrayList would guarantee this wasn't necessary if we took an iterator but because we're using a classic for-loop I believe we could still be `< referenceChangeListeners.size()` three lines ago and get an OOB exception when we go to get it.